pkg/attribute/protoBag.go

```diff
@@ -50,9 +64,56 @@ func (pb *ProtoBag) Get(name string) (interface{}, bool) {
 	index, ok := pb.getIndex(name)
 	if !ok {
 		// the string is not in the dictionary, and hence the attribute is not in the proto either
+		pb.trackReference(name, mixerpb.ABSENCE)
 		return nil, false
 	}
 
+	result, ok := pb.internalGet(name, index)
+	if !ok {
+		// the named attribute was not present
+		pb.trackReference(name, mixerpb.ABSENCE)
+		return nil, false
+	}
+
+	pb.trackReference(name, mixerpb.EXACT)
+	return result, ok
+}
+
+// Get the set of attributes that have been referenced through this bag.
+func (pb *ProtoBag) GetReferencedAttributes() mixerpb.ReferencedAttributes {
+	output := mixerpb.ReferencedAttributes{}
+
+	ds := newDictState(nil)
+
+	output.AttributeMatches = make([]mixerpb.ReferencedAttributes_AttributeMatch, len(pb.referencedAttrs))
+	i := 0
+	for k, v := range pb.referencedAttrs {
+		output.AttributeMatches[i] = mixerpb.ReferencedAttributes_AttributeMatch{
+			Name:      ds.getIndex(k),
```

Comment:
Oh wow. It took me a while to understand what this is doing. I'm not a fan of `getIndex` mutating the underlying store, I think that kind of mutation might bite us down the road. In Java I'd call this method `getGlobalIndexOrCreateMessageIndex`. Maybe change this to `synthesizeIndex` or `getOrCreateIndex`, or similar, to denote that this is _not_ just a lookup in the dicts that the dictState was created with.